confidant/services/certificatemanager.py

```diff
@@ -0,0 +1,397 @@
+        ]
+        if self.settings['csr_country_name']:
+            name_attributes.append(
+                x509.NameAttribute(
+                    NameOID.COUNTRY_NAME,
+                    self.settings['csr_country_name'],
+                )
+            )
+        if self.settings['csr_state_or_province_name']:
+            name_attributes.append(
+                x509.NameAttribute(
+                    NameOID.STATE_OR_PROVINCE_NAME,
+                    self.settings['csr_state_or_province_name'],
+                )
+            )
+        if self.settings['csr_locality_name']:
+            name_attributes.append(
+                x509.NameAttribute(
+                    NameOID.LOCALITY_NAME,
+                    self.settings['csr_locality_name'],
+                )
+            )
+        if self.settings['csr_organization_name']:
+            name_attributes.append(
+                x509.NameAttribute(
+                    NameOID.ORGANIZATION_NAME,
+                    self.settings['csr_organization_name'],
+                )
+            )
+        return x509.Name(name_attributes)
+
+    def generate_csr(self, key, cn, san=None):
+        """
+        Using the provided rsa key object, a string common name, and a list of
+        string subject alternative names, generate and return a csr object.
+        """
+        csr = x509.CertificateSigningRequestBuilder().subject_name(
+            self.generate_x509_name(cn)
+        )
+        if san:
+            dns_names = self.encode_san_dns_names(san)
+            csr = csr.add_extension(
+                x509.SubjectAlternativeName(dns_names),
+                critical=False,
+            )
+        return csr.sign(key, hashes.SHA256(), default_backend())
+
+    def encode_csr(self, csr):
+        """
+        Return a PEM string encoded version of the csr object.
+        """
+        return csr.public_bytes(serialization.Encoding.PEM)
+
+    def decode_csr(self, pem_csr):
+        """
+        Return a csr object from the pem encoded csr.
+        """
+        return x509.load_pem_x509_csr(pem_csr, default_backend())
+
+    def get_csr_common_name(self, csr):
+        """
+        From the provided csr object, return the string value of the common
+        name attribute.
+        """
+        cns = csr.subject.get_attributes_for_oid(NameOID.COMMON_NAME)
+        if cns:
+            # get_attributes_for_oid returns a list, but there should only be a
+            # single cn attribute, so just return the first item.
+            return cns[0].value
+        return None
+
+    def get_csr_san(self, csr):
+        """
+        From the provided csr object, return a list of the string values of the
+        subjust alternative name extension.
+        """
+        dns_names = []
+        try:
+            san = csr.extensions.get_extension_for_class(
+                x509.SubjectAlternativeName
+            )
+        except ExtensionNotFound:
+            san = []
```

Comment:
nit:  I think better to set this to `None` instead of an empty list?    I'm interpreting the return to be a list, but looks like it's a single object where `san.value` is the list.

Reply:
Ah. Good catch. Will change.